### PR TITLE
Nav task improvements

### PR DESCRIPTION
**Pull request by danielgordon10**
## Motivation and Context

Adding the top-down-map measure and the static pointnav measure. Part of this change draws a line using opencv which is a dependency you might not want to add, but it's pip installable.

## How Has This Been Tested

Making ICCV videos.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: habitat/tasks/nav/nav_task.py [hunk 1]
```diff
@@ -6,9 +6,11 @@
 
 from typing import Any, List, Optional, Type
 
-import habitat
+import cv2
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: Can you update the `requirements.txt` as well as `setup.py` and add `cv2` dependency in there, I think it is okay to have a dependency on `cv2` now as we are supporting top-down maps.

File: habitat/tasks/nav/nav_task.py [hunk 2]
```diff
@@ -204,6 +207,84 @@ def get_observation(self, observations, episode):
...
+            sensor_shape = (2,)
+        return spaces.Box(
+            low=np.finfo(np.float32).min,
+            high=np.finfo(np.float32).max,
+            shape=sensor_shape,
+            dtype=np.float32,
+        )
+
+    def get_observation(self, observations, episode):
+        episode_unique_id = (episode.episode_id, episode.scene_id)
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: replace `episode_unique_id` with `episode_id` and `self.episode_unique_id` with `self.current_episode_id`

File: habitat/tasks/nav/nav_task.py [hunk 3]
```diff
@@ -363,6 +443,148 @@ def update_metric(self, episode, action):
...
+
+    def __init__(self, sim: Simulator, config: Config):
+        self._sim = sim
+        self._draw_source_and_target = config.DRAW_SOURCE_AND_TARGET
+        self._draw_border = config.DRAW_BORDER
+        self._grid_delta = 3
+        self._step_count = 0
+        self._max_steps = config.MAX_EPISODE_STEPS
+        self._map_resolution = (config.MAP_RESOLUTION, config.MAP_RESOLUTION)
+        self._num_samples = 20000
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: Move this inside top down map config.

File: habitat/tasks/nav/nav_task.py [hunk 4]
```diff
@@ -363,6 +443,148 @@ def update_metric(self, episode, action):
...
+            self._coordinate_max - self._coordinate_min
+        ) / self._map_resolution[0]
+        super().__init__()
+
+    def _get_uuid(self, *args: Any, **kwargs: Any):
+        return "top_down_map"
+
+    def _check_valid_nav_point(self, point: List[float]):
+        return (
+            0.01
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: Use class parameters for constants.

File: habitat/tasks/nav/nav_task.py [hunk 5]
```diff
@@ -363,6 +443,148 @@ def update_metric(self, episode, action):
...
+            < self._sim.geodesic_distance(
+                point, [point[0], point[1] + 0.1, point[2]]
+            )
+            < 0.11
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: Use class parameters for constants.

File: habitat/tasks/nav/nav_task.py [hunk 6]
```diff
@@ -363,6 +443,148 @@ def update_metric(self, episode, action):
...
+        if self._draw_source_and_target:
+            # mark source point
+            s_x, s_y = maps.to_grid(
+                episode.start_position[0],
+                episode.start_position[2],
+                self._coordinate_min,
+                self._coordinate_max,
+                self._map_resolution,
+            )
+            point_padding = 2 * int(np.ceil(self._map_resolution[0] / 1250))
```
Comment (abhiskk) [inline habitat/tasks/nav/nav_task.py]: Significane of 1250 here? Maybe you can use the value of resolution directly.

File: habitat/tasks/nav/nav_task.py (hunk 1 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: It is in there. "opencv-python>=3.3.0" is what it's called. But I moved it to before the comment about visualization if that makes you happier.

File: habitat/tasks/nav/nav_task.py (hunk 5 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: @erikwijmans Was there a new habitat-sim function that we are supposed to use now for checking whether a point is valid or not?

File: habitat/tasks/nav/nav_task.py (hunk 6 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: 1250 makes it invariant to the map_resolution changing in terms of the final image. It was just a value that looked pretty good. I can parameterize it.

File: habitat/tasks/nav/nav_task.py (hunk 3 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: What do you mean by this? Like move it into the defaults.py config somewhere?

File: habitat/tasks/nav/nav_task.py (hunk 5 above)
Comment (erikwijmans) [reply habitat/tasks/nav/nav_task.py]: Yep, https://github.com/facebookresearch/habitat-api/blob/master/habitat/sims/habitat_simulator.py#L341

File: habitat/config/default.py [hunk 7]
```diff
@@ -54,6 +54,15 @@
 _C.TASK.SPL.TYPE = "SPL"
 _C.TASK.SPL.SUCCESS_DISTANCE = 0.2
 # -----------------------------------------------------------------------------
+# # TopDownMap MEASUREMENT
+# -----------------------------------------------------------------------------
+_C.TASK.TOP_DOWN_MAP = CN()
+_C.TASK.TOP_DOWN_MAP.TYPE = "TOP_DOWN_MAP"
```
Comment (erikwijmans) [inline habitat/config/default.py]:
```suggestion
_C.TASK.TOP_DOWN_MAP.TYPE = "TopDownMap"
```

File: habitat/tasks/nav/nav_task.py [hunk 8]
```diff
@@ -363,6 +446,160 @@ def update_metric(self, episode, action):
...
+                self._map_resolution,
+            )
+            top_down_map[
+                t_x - point_padding : t_x + point_padding + 1,
+                t_y - point_padding : t_y + point_padding + 1,
+            ] = 6
+
+        return top_down_map
+
+    def reset_metric(self, episode):
```
Comment (erikwijmans) [inline habitat/tasks/nav/nav_task.py]: Needs to zero out the step_count here.

`self._step_count = 0`

Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: Good catch

File: habitat/tasks/nav/nav_task.py [hunk 9]
```diff
@@ -363,6 +444,147 @@ def update_metric(self, episode, action):
...
+        house_map = house_map[
+            self._ind_x_min
+            - self._grid_delta : self._ind_x_max
+            + self._grid_delta,
+            self._ind_y_min
+            - self._grid_delta : self._ind_y_max
+            + self._grid_delta,
+        ]
+
+        self._metric = house_map
```
Comment (mathfac) [inline habitat/tasks/nav/nav_task.py]: In `habitat.Measure` metric is a number. Maybe that's good to switch it to something more general.

File: habitat/tasks/nav/nav_task.py (hunk 3 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: Fixed? I'm still not really understanding what you guys want.

File: habitat/tasks/nav/nav_task.py [hunk 10]
```diff
@@ -204,6 +208,84 @@ def get_observation(self, observations, episode):
...
+            in cartesian or polar coordinates.
+    """
+
+    def __init__(self, sim, config):
+        self._sim = sim
+        self._goal_format = getattr(config, "GOAL_FORMAT", "CARTESIAN")
+        assert self._goal_format in ["CARTESIAN", "POLAR"]
+
+        super().__init__(sim, config)
+        self.initial_vector = None
```
[the remark this answers is not in the document]
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: Done

File: habitat/tasks/nav/nav_task.py (hunk 9 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: I get what you mean, but in the interface it doesn't seem to specify the type. What if we made this a Sensor or an Info?

File: habitat/tasks/nav/nav_task.py (hunk 3 above)
Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: We want all consts to be part of config:
`self._num_samples = 20000 ` => `_C.TASK.TOP_DOWN_MAP.NUM_SAMPLES = 20000`.
Same for `_grid_delta`.

File: habitat/tasks/nav/nav_task.py (hunk 9 above)
Comment (mathfac) [reply habitat/tasks/nav/nav_task.py]: > I get what you mean, but in the interface it doesn't seem to specify the type. What if we made the top down map a sensor or an Info?
My bad, ignore this comment please.

File: habitat/tasks/nav/nav_task.py (hunk 3 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: I see. Done

File: habitat/tasks/nav/nav_task.py [hunk 11]
```diff
@@ -204,6 +208,78 @@ def get_observation(self, observations, episode):
...
+        config: config for the PointGoal sensor. Can contain field for
+            GOAL_FORMAT which can be used to specify the format in which
+            the pointgoal is specified. Current options for goal format are
+            cartesian and polar.
+    Attributes:
+        _goal_format: format for specifying the goal which can be done
+            in cartesian or polar coordinates.
+    """
+
+    def __init__(self, sim, config):
```
Comment (mathfac) [inline habitat/tasks/nav/nav_task.py]:
```suggestion
    def __init__(self, sim: Simulator, config: Config):
```

File: habitat/tasks/nav/nav_task.py [hunk 12]
```diff
@@ -363,6 +438,152 @@ def update_metric(self, episode, action):
...
+
+        range_x = np.where(np.any(top_down_map, axis=1))[0]
+        range_y = np.where(np.any(top_down_map, axis=0))[0]
+
+        self._ind_x_min = range_x[0]
+        self._ind_x_max = range_x[-1]
+        self._ind_y_min = range_y[0]
+        self._ind_y_max = range_y[-1]
+
+        MAP_SOURCE_POINT_INDICATOR = 4
```
Comment (mathfac) [inline habitat/tasks/nav/nav_task.py]: Why not reuse `maps.MAP_SOURCE_POINT_INDICATOR`?

Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: ...whoops, didn't mean to leave that in.

File: habitat/tasks/nav/nav_task.py (hunk 11 above)
Comment (danielgordon10) [reply habitat/tasks/nav/nav_task.py]: Done

**Pull request merged**